helm provisioner initial drop (#473)
operator-framework/rukpak: internal/provisioner/helm/README.md

# Plain Provisioner

## Summary

> The `helm` provisioner is able to instantiate a given `helm+v0` bundle with a specified helm chart onto a cluster and then install the helm chart,

Reviewer: This brings up the interesting point of whether we need the `+v0` -- since helm is an external format, I'm not sure we are going to have the same metadata questions we have regarding our own, internal, `plain` format. I'm personally for dropping the v0 and just calling them helm bundles in the short term. 

This would create less confusion on what helm bundles are and how to use them.

Author: OK.  That makes sense.

Reviewer 2: I'd probably call them helm v3 bundles. Helm made some minor (maybe breaking, can't remember?) changes between v2 and v3.

Author: "v3" makes sense.  I'll put "+v3".

in the cluster. It does so by reconciling `Bundle` and `BundleDeployment` types that have
the `spec.provisionerClassName` field set to `core.rukpak.io/helm`. This field must be set to the correct provisioner
name in order for the `helm` provisioner to see and interact with the bundle.

### Install and apply a specific version of a `helm+v0` bundle

> > :warning: Anyone with the ability to create or update BundleDeployment objects can become cluster admin. It's important

Reviewer 3: nit:
```suggestion
> :warning: Anyone with the ability to create or update BundleDeployment objects can become a cluster admin. It's important
```

Author: Thanks!

> to limit access to this API via RBAC to only those that explicitly require access, as well as audit your bundles to
> ensure the content being installed on-cluster is as-expected and secure.

The `helm` provisioner can install and make available a specific `helm+v0` bundle in the cluster.

Simply create a `BundleDeployment` resource that contains the desired specification of a Bundle resource.

> The `helm` provisioner will retrive the specifie helm chart and instantiate Bundle onto the cluster, and eventually install the helm chart

Reviewer: ```suggestion
The `helm` provisioner will retrieve the specific helm chart and instantiate the Bundle onto the cluster, and eventually install the helm chart
```

Author: Thanks!

 on the cluster.

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: BundleDeployment
metadata:
  name: my-ahoy
spec:
  provisionerClassName: core.rukpak.io/helm
  config:
    valuesFrom:
      configMapName: my-ahoy-values
  template:
    metadata:
      labels:
        app: my-ahoy
    spec:
      provisionerClassName: core.rukpak.io/helm
      source:
        http:
          url: https://github.com/helm/examples/releases/download/hello-world-0.1.0/hello-world-0.1.0.tgz  
        type: http
```

> For the helm chart, the values file can be applied during the installtation of the chart.  The values.yaml has to be in a ConfigMap in the namespace where the helm provisioner is running.

Reviewer 3: nit:
```suggestion
For the helm chart, the values file can be applied during the installtation of the chart. The values.yaml has to be in a ConfigMap in the namespace where the helm provisioner is running.
```

Author: Thanks!

The ConfigMap can be created by the kubectl

```console

> kubectl create configmap my-ahoy-values --from-file=values.yaml

Reviewer: Could we show what is actually in the `values.yaml` file? Might be helpful for users to see.

Author: OK.  I'll add it.  Here is the values.yaml file from a sample.  I'll change the sample from apache to helm/examples to make it easy.

```
# Default values for hello-world.
# This is a YAML-formatted file.
# Declare variables to be passed into your templates.

replicaCount: 1

image:
  repository: nginx
  pullPolicy: IfNotPresent
  # Overrides the image tag whose default is the chart appVersion.
  tag: ""

nameOverride: "formvalues"
fullnameOverride: ""

serviceAccount:
  # Specifies whether a service account should be created
  create: true
  # Annotations to add to the service account
  annotations: {}
  # The name of the service account to use.
  # If not set and create is true, a name is generated using the fullname template
  name: ""

service:
  type: ClusterIP
  port: 80

```

```

The ConfigMap name is set in the BundleDeployment

```yaml
  config:
    valuesFrom:
      configMapName: my-ahoy-values
```

> Note: the generated Bundle will contain the BundleDeployment's metadata.Name as a prefix, followed by
> the hash of the provided template.

> As the bundle content is retrived and stored onto the cluster via the defined storage mechanism, the bundle status

Reviewer: ```suggestion
As the bundle content is retrieved and stored onto the cluster via the defined storage mechanism, the bundle status
```

Author: Thanks!

will be updated to Unpacked, indicating that all its contents have been stored on-cluster.

```console
$ kubectl get bundle -l app=my-ahoy
NAME                 TYPE   PHASE      AGE

> my-ahoy-5764594dc8   http   Unpacked   33s

Reviewer: should this be `TYPE` helm?

Author: This column shows the source type.  We can add provisioner class name.

      labels:
        app: my-ahoy
    spec:
      provisionerClassName: core.rukpak.io/helm
      source:
        http:
          url: https://github.com/helm/examples/releases/download/hello-world-0.1.0/hello-world-0.1.0.tgz  
        type: http
```

### Deleting the sample chart and Local Kind Cluster

To clean up from the installation, simply remove the BundleDeployment from the cluster. This will remove all references
resources including the deployment, RBAC, and the operator namespace.

> Note: There's no need to manually clean up the Bundles that were generated from a BundleDeployment resource. The plain provisioner places owner references on any Bundle that's generated from an individual BundleDeployment resource.

```bash
# Delete the combo BundleDeployment
kubectl delete bundledeployments.core.rukpak.io my-ahoy
```

A message should show that the BundleDeployment was deleted and now the cluster state is the same as it was
prior to installing the operator.

```console
$ kubectl delete bundledeployments.core.rukpak.io my-ahoy
bundledeployment.core.rukpak.io "my-ahoy" deleted
```

To stop and clean up the kind cluster, delete it:

```bash
# Clean up kind cluster

> make kind-cluster-cleanup

Reviewer: I think `kind delete cluster` also makes sense, if we want to leave the host in the same exact state as before running the cluster.

Author: OK.  The `make run` creates the rukpak cluster.  So I'll change the line to `kind delete clusters rukpak`

kind: BundleDeployment
metadata:
  name: my-ahoy
spec:
  provisionerClassName: core.rukpak.io/helm
  config:
    values: |
      # Default values for hello-world.
      # This is a YAML-formatted file.
      # Declare variables to be passed into your templates.
      replicaCount: 1
      image:
        repository: nginx
        pullPolicy: IfNotPresent
        # Overrides the image tag whose default is the chart appVersion.
        tag: ""
      nameOverride: "formvalues"
      fullnameOverride: ""
      serviceAccount:
        # Specifies whether a service account should be created
        create: true
        # Annotations to add to the service account
        annotations: {}
        # The name of the service account to use.
        # If not set and create is true, a name is generated using the fullname template
        name: ""
      service:
        type: ClusterIP
        port: 80
  template:
    metadata:
      labels:
        app: my-ahoy
    spec:
      provisionerClassName: core.rukpak.io/helm
      source:
        https:
          url: https://github.com/helm/examples/releases/download/hello-world-0.1.0/hello-world-0.1.0.tgz  
        type: https
```

> For the helm chart, the values file embedded in the `confg` can be applied during the installtation of the chart.

Reviewer 2: ```suggestion
For the helm chart, the values file embedded in the `config` can be applied during the installation of the chart.
```

Author: Thanks.

> # Plain Provisioner

Reviewer 2: ```suggestion
# Helm Provisioner
```

Author: Thanks!

the `spec.provisionerClassName` field set to `core.rukpak.io/helm`. This field must be set to the correct provisioner
name in order for the `helm` provisioner to see and interact with the bundle.

### Install and apply a specific version of a `helm+v3` bundle

> :warning: Anyone with the ability to create or update BundleDeployment objects can become a cluster admin. It's important
> to limit access to this API via RBAC to only those that explicitly require access, as well as audit your bundles to
> ensure the content being installed on-cluster is as-expected and secure.

The `helm` provisioner can install and make available a specific `helm+v3` bundle in the cluster.

Simply create a `BundleDeployment` resource that contains the desired specification of a Bundle resource.
The `helm` provisioner will retrieve the specific helm chart and instantiate the Bundle onto the cluster, and eventually install the helm chart
 on the cluster.

```yaml
apiVersion: core.rukpak.io/v1alpha1
kind: BundleDeployment
metadata:
  name: my-ahoy
spec:
  provisionerClassName: core.rukpak.io/helm
  config:
    valuesFrom:

>       configMapName: my-ahoy-values

Reviewer 2: WDYT about (for now at least) embedding the values directly in the bundle deployment either as a `map[string]interface{}`, or as a `string`, rather than referencing a configmap? That would decrease the complexity (no need to handle all the possibilities around a configmap being present/not present/showing up/disappearing/etc). 

E.g.

```yaml
spec:
  config:
    values: # map[string]interface{}
      replicaCount: 1
      
      image:
        repository: nginx
        pullPolicy: IfNotPresent
        # Overrides the image tag whose default is the chart appVersion.
        tag: ""
    # or
    values: |  # string
      replicaCount: 1
      
      image:
        repository: nginx
        pullPolicy: IfNotPresent
        # Overrides the image tag whose default is the chart appVersion.
        tag: ""
```

Author: I see.  It seems better.  I'll change to the `string` one.

Author: done.
